mypy/checkexpr.py

```diff
@@ -3108,112 +3108,134 @@ def infer_lambda_type_using_context(self, e: LambdaExpr) -> Tuple[Optional[Calla
+        if type_info is None:
+            self.chk.fail(message_registry.UNSUPPORTED_ARG_1_FOR_SUPER, e)
+            return AnyType(TypeOfAny.from_error)
+
+        instance_info = self._type_info(instance_type)
+        if instance_info is None:
+            self.chk.fail(message_registry.UNSUPPORTED_ARG_2_FOR_SUPER, e)
+            return AnyType(TypeOfAny.from_error)
+
+        mro = instance_info.mro
+
+        # The base is the first MRO entry *after* type_info that has a member
+        # with the right name
+        try:
+            index = mro.index(type_info)
+        except ValueError:
+            self.chk.fail(message_registry.SUPER_ARG_2_NOT_INSTANCE_OF_ARG_1, e)
+            return AnyType(TypeOfAny.from_error)
+
+        for base in mro[index+1:]:
+            if e.name in base.names or base == mro[-1]:
+                if e.info and e.info.fallback_to_any and base == mro[-1]:
+                    # There's an undefined base class, and we're at the end of the
+                    # chain.  That's not an error.
+                    return AnyType(TypeOfAny.special_form)
+
+                return analyze_member_access(name=e.name,
+                                             typ=instance_type,
+                                             is_lvalue=False,
+                                             is_super=True,
+                                             is_operator=False,
+                                             original_type=instance_type,
+                                             override_info=base,
+                                             context=e,
+                                             msg=self.msg,
+                                             chk=self.chk,
+                                             in_literal_context=self.is_literal_context())
+
+        assert False, 'unreachable'
+
+    def _super_arg_types(self, e: SuperExpr) -> Union[Type, Tuple[Type, Type]]:
+        if not e.info:
+            # Invalid super.  This has been reported by the semantic analyzer.
+            return AnyType(TypeOfAny.from_error)
+        elif not self.chk.in_checked_function():
+            return AnyType(TypeOfAny.unannotated)
+
+        current_instance = fill_typevars(e.info)
+        current_type = TypeType(current_instance)
+
+        instance_type = current_instance  # type: Type
+        type_type = current_type  # type: Type
+
+        if len(e.call.args) == 0:
+            if self.chk.options.python_version[0] == 2:
+                self.chk.fail(message_registry.TOO_FEW_ARGS_FOR_SUPER, e)
+                return AnyType(TypeOfAny.from_error)
+            elif self.chk.scope.active_class():
+                self.chk.fail(message_registry.SUPER_OUTSIDE_OF_METHOD_NOT_SUPPORTED, e)
+                return AnyType(TypeOfAny.from_error)
+
+            # Use the type of the self argument, in case it was annotated
+            method = self.chk.scope.top_function()
```

Comment:
What if this was a nested function within a method?

Reply:
Also copied from the old code, but I wouldn't be surprised if this is wrong in that case.  I guess we need to walk the scopes to find the current method?

Reply:
Oh actually I think this code is right.  The Python interpreter uses the innermost function even if it's not a method.  Something like

```
class A:
    def f(self): pass

class B(A):
    def f(self):
        def g():
            super().f()
        g()

B().f()
```

gives

```
RuntimeError: super(): no arguments
```

`testSuperInMethodWithNoArguments` already checks for this.